### PR TITLE
Add the Collatz Conjecture as a Benchmark

### PR DESCRIPTION
**Pull request by Mr4k**
Added a new benchmark which is the sum of the first n collatz numbers. It's a fun recursive problem which should benefit from memoization but not as directly as fib and is a little more complex.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,2 +1,7 @@
 bench:
 	cd benchmark && go test -bench=. -benchtime=10s
+fmt:
```
Comment (Mr4k) [inline Makefile:3]: Easy of use commands b/c I always forget go

File: benchmark/benchmark_test.go [hunk 2]
```diff
@@ -37,3 +37,21 @@ func BenchmarkPureFib(b *testing.B) {
 	}
 	result = r
 }
+
+func BenchmarkCollatz(b *testing.B) {
```
Comment (Mr4k) [inline benchmark/benchmark_test.go:35]: This benchmark sums up the first 100 [collatz numbers](https://en.wikipedia.org/wiki/Collatz_conjecture). 

Results on my computer are as follows:
```
BenchmarkCollatz-4       	    2750	   4459093 ns/op
BenchmarkPureCollatz-4   	   12122	    963438 ns/op
```

File: evaluator/evaluator.go [hunk 3]
```diff
@@ -258,6 +259,8 @@ func evalFloatInfixExpression(operator string, left object.Object, right object.
 		return nativeBoolToBooleanObject(lVal == rVal)
 	case "!=":
 		return nativeBoolToBooleanObject(lVal != rVal)
+	case "%":
```
Comment (Mr4k) [inline evaluator/evaluator.go:262]: I hit a strange error where dividing a number turns it into a Float which is no longer valid for modulus. I just decided to add modulus for floats but lmk if you think this is a good idea.

File: parser/parser_test.go [hunk 4]
```diff
@@ -12,7 +12,7 @@ func TestComments(t *testing.T) {
 	let x = 5; // comment
 	// another comment
 	1 + 2
-	`;
```
Comment (Mr4k) [inline parser/parser_test.go:15]: no changes to this file but for some reason go fmt doesn't like these couple lines. Might as well appease it?

File: evaluator/evaluator.go (hunk 3 above)
Comment (Mr4k) [reply evaluator/evaluator.go:262]: Related: I think, the error system might be a little broken right now (errors are not getting piped up) but I think that should be a separate pr

File: benchmark/benchmark_test.go [hunk 5]
```diff
@@ -37,3 +37,21 @@ func BenchmarkPureFib(b *testing.B) {
...
+	var r object.Object
+	program, env := testBuild(`let collatz = fn(n) { if (n==1) { 0 } else { if (n%2 == 0) { collatz(n/2) + 1 } else { collatz(3*n + 1) + 1 }}};let compute_sum_of_first_n_collatz = fn(n) { if (n== 1) { 0 } else { collatz(n) + compute_sum_of_first_n_collatz(n - 1) }}; compute_sum_of_first_n_collatz(100)`)
```
Comment (jemmaissroff) [inline benchmark/benchmark_test.go]: Would it make sense to split this into a multi line string?

File: benchmark/benchmark_test.go [hunk 6]
```diff
@@ -37,3 +37,21 @@ func BenchmarkPureFib(b *testing.B) {
...
+	for i := 0; i < b.N; i++ {
+		r = evaluator.Eval(program, env)
+	}
+	result = r
```
Comment (jemmaissroff) [inline benchmark/benchmark_test.go]: Did we ever end up validating if we need this variable assignment for accurate benchmarking?

File: benchmark/benchmark_test.go [hunk 7]
```diff
@@ -37,3 +37,21 @@ func BenchmarkPureFib(b *testing.B) {
...
+}
+
+func BenchmarkPureCollatz(b *testing.B) {
+	var r object.Object
+	program, env := testBuild(`let collatz = pfn(n) { if (n==1) { 0 } else { if (n%2 == 0) { collatz(n/2) + 1 } else { collatz(3*n + 1) + 1 }}};let compute_sum_of_first_n_collatz = fn(n) { if (n== 1) { 0 } else { collatz(n) + compute_sum_of_first_n_collatz(n - 1) }}; compute_sum_of_first_n_collatz(100)`)
```
Comment (jemmaissroff) [inline benchmark/benchmark_test.go]: Same as above re multiline

File: evaluator/evaluator.go (hunk 3 above)
Comment (jemmaissroff) [reply evaluator/evaluator.go:262]: This is awesome, and thanks for making the other PR too!

Totally down to have modulus for floats too! Another way we could do this is cast to an int above, so `collatz(int(n/2))` _or_ `if (int(n) % 2 == 0)`

**Review by jemmaissroff — CHANGES_REQUESTED** [review]

File: evaluator/evaluator.go (hunk 3 above)
Comment (Mr4k) [reply evaluator/evaluator.go:262]: oh did not know there were casts! I feel like float modulus is useful(?) so might merge that part in

File: benchmark/benchmark_test.go (hunk 5 above)
Comment (Mr4k) [reply benchmark/benchmark_test.go]: yea sounds good

File: benchmark/benchmark_test.go (hunk 6 above)
Comment (Mr4k) [reply benchmark/benchmark_test.go]: hmm yea we can toss this

File: benchmark/benchmark_test.go (hunk 7 above)
Comment (Mr4k) [reply benchmark/benchmark_test.go]: 👍

**Pull request merged**